Support any dictionary types JsonExtensionData

Romanx opened the pull request:
This changes the way that ExtensionData is gathered and assigned so that properties are added and assigned before finishing parsing an object.

This works for the following dictionary types
- `Dictionary<,>`
- `SortedDictionary<,>`
- `IDictionary<,>`
- `IReadOnlyDictionary<,>`
- `ImmutableDictionary<,>`
- `ImmutableSortedDictionary<,>`
- `IImmutableDictionary<,>`
- `ConcurrentDictionary<,>`

closes #31645

Romanx committed:
Adds tests for all supported extension data types

Romanx committed:
Adds an internal convert method on JsonConverter

This is convert one type to another using the internal cached delegate
the converter has.

Romanx committed:
Added special casing for Readonly and Immutables

This allows those types to be valid as DataExtensions and are special
cased since others can implement those interfaces and that would break
conversion

Romanx committed:
ExtensionData is stored on frame and assigned

This is converted and assigned when finishing an object and uses the
ExtensionData converter to create the correct types. For Immutable types
this is different since we create a standard dictionary and convert
before assignment

dnfclas commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=35158) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=35158) before we can accept your contribution.<br/><br/>:x: Romanx [sign now](https://cla.dotnetfoundation.org/dotnet/runtime?pullRequest=35158)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/dotnet/runtime?pullRequest=35158) it.</sub>

layomia commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/ImmutableDictionaryOfStringTValueConverter.cs:
```diff
@@ -31,6 +31,11 @@ protected override void ConvertCollection(ref ReadStack state, JsonSerializerOpt
             state.Current.ReturnValue = GetCreatorDelegate(options)((Dictionary<string, TValue>)state.Current.ReturnValue!);
         }
 
+        internal override object Convert(object value, JsonSerializerOptions options)
```
```suggestion
        internal override object ConvertToDictionary(object value, JsonSerializerOptions options)
```

This needs to be more descriptive.

layomia commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs:
```diff
@@ -120,19 +120,23 @@ public static partial class JsonSerializer
         }
 
         internal static void CreateDataExtensionProperty(
```
The reflection being added to this method will occur during deserialization (as opposed to during the serializer's warm-up, when `JsonClassInfo`s and `JsonPropertyInfo`s are created), and will lead to bad throughput performance.

I'll come back to this in a follow-up review, but I think the type metadata being retrieved here can be cached on the `JsonClassInfo` for the dictionary type during warm-up.

layomia commented on src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializer.Read.HandlePropertyName.cs:
```diff
@@ -142,11 +146,20 @@ public static partial class JsonSerializer
+                        ThrowHelper.ThrowNotSupportedException_SerializationNotSupported(jsonPropertyInfo.DeclaredPropertyType);
+                    }
+
+                    extensionData = Activator.CreateInstance(typeof(Dictionary<,>).MakeGenericType(genericArgs));
```
We should be able to add a new abstract method on `JsonPropertyInfo`, and leverage the generic support on `JsonProperty<T>` to to create this instance. The call could look like this:

```
extensionData = jsonPropertyInfo.RuntimeClassInfo.ElementClassInfo.PropertyInfoForClassInfo.CreateDictionaryOfStringTValueInstance();
```

layomia reviewed (commented):
Thanks for the PR! I took a quick pass and will review again shortly.

This PR adds new state to `ReadStackFrame` which is perf sensitive, so we'll want to get before/after runs for the [System.Text.Json `ReadJson` and `WriteJson` benchmarks](https://github.com/dotnet/performance/tree/master/src/benchmarks/micro/libraries/System.Text.Json/Serializer) in the performance repo, to validate these changes. You can see https://github.com/dotnet/performance/blob/master/docs/benchmarking-workflow-dotnet-runtime.md#preventing-regressions for how to get started.

We'll also need benchmarks specific to deserializing objects with extension data that cover existing functionality, e.g. deserializing 10-15 extra JSON property values into extension data, along with properties that map to CLR object properties.

I can run these benchmarks if you can't get around to them.

layomia commented on src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs:
```diff
@@ -1066,5 +1066,53 @@ public override void Write(Utf8JsonWriter writer, JsonElement value, JsonSeriali
+        }
+
+        [Theory]
+        [InlineData(typeof(HasExtensionData<Dictionary<string, object>>))]
```
We need negative tests where the generic dictionary types are supported, but they key type is not `string` and/or the value type is not `object` or `JsonElement`.

Romanx replied:
The negative tests we have as `ExtensionProperty_InvalidDictionary` but we can add tests for the specific types if you think that's worthwhile.

Do you mean non generic dictionaries like `IDictionary`? I don't think that will work since it needs to match either `<string, object>` or `<string, JsonElement>`.

layomia commented on src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs:
```diff
@@ -903,8 +905,6 @@ public static void DeserializeIntoImmutableDictionaryProperty()
             JsonSerializer.Deserialize<ClassWithExtensionPropertyPrivateConstructor>(@"{}");
             JsonSerializer.Deserialize<ClassWithExtensionPropertyPrivateConstructorJsonElement>(@"{}");
 
-            Assert.Throws<NotSupportedException>(() => JsonSerializer.Deserialize<ClassWithExtensionPropertyAsImmutable>("{\"hello\":\"world\"}"));
```
Please add JSON properties to the above tests and verify the results of the deserialization operations.

Romanx replied:
I think the tests above are to verify that no exceptions are thrown when the extension data isn't used but does when it tries to deserialize into it. I removed these two since they're no longer accurate and it does serialize

Romanx commented on src/libraries/System.Text.Json/tests/Serialization/ExtensionDataTests.cs:
```diff
@@ -1066,5 +1066,53 @@ public override void Write(Utf8JsonWriter writer, JsonElement value, JsonSeriali
+        [InlineData(typeof(HasExtensionData<ImmutableSortedDictionary<string, object>>))]
+        [InlineData(typeof(HasExtensionData<IImmutableDictionary<string, object>>))]
+        [InlineData(typeof(HasExtensionData<ConcurrentDictionary<string, object>>))]
+        public static void TypeValidityAsExtensionDataWithObjectValue(Type type)
```
The test for converters are handled separately from these additional tests such as the test `CustomObjectConverterInExtensionProperty`. We could add more tests for the specific types here but that wouldn't be testing any new codepaths since there's no specific logic for converters changed here so these verify we didn't break anything existing

Romanx committed:
Apply changes suggested by @layomia for reflection

This caches type metadata to avoid reflection overhead possibly reducing
throughput

Romanx committed:
Use expression body for throw only

Romanx commented:
@layomia I had a look at your suggestions and made a commit merging them along with rebasing all the changes on top of master. I also made some replies on your comments about some cases that are already covered by other tests

I tried running the performance tests using master as a baseline (without any specific data extension tests added) and got the following using a 2% threshold as suggested in the doc:

summary:
better: 7, geomean: 1.042
worse: 16, geomean: 1.035
total diff: 23

| Slower                                                                           | diff/base | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.ReadJson<LoginViewModel>.DeserializeFromStr |      1.14 |           614.24 |           698.81 |         |
| System.Text.Json.Serialization.Tests.WriteJson<ArrayList>.SerializeObjectPropert |      1.04 |         14160.18 |         14768.43 |         |
| System.Text.Json.Serialization.Tests.ReadJson<MyEventsListerViewModel>.Deseriali |      1.04 |        452811.41 |        471336.04 |         |
| System.Text.Json.Serialization.Tests.WriteJson<LoginViewModel>.SerializeToStream |      1.04 |           550.35 |           571.60 |         |
| System.Text.Json.Serialization.Tests.ReadJson<LoginViewModel>.DeserializeFromUtf |      1.03 |           560.35 |           577.89 |         |
| System.Text.Json.Serialization.Tests.WriteJson<Int32>.SerializeToUtf8Bytes       |      1.03 |           175.85 |           180.79 |         |
| System.Text.Json.Serialization.Tests.WriteJson<BinaryData>.SerializeToString     |      1.03 |           761.69 |           782.26 |         |
| System.Text.Json.Serialization.Tests.WriteJson<BinaryData>.SerializeObjectProper |      1.03 |           997.93 |          1024.76 |         |
| System.Text.Json.Serialization.Tests.ReadJson<HashSet<String>>.DeserializeFromSt |      1.03 |         18433.70 |         18926.40 |         |
| System.Text.Json.Serialization.Tests.WriteJson<Location>.SerializeToString       |      1.03 |          1080.45 |          1109.17 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ArrayList>.DeserializeFromString   |      1.03 |         59883.01 |         61418.66 |         |
| System.Text.Json.Serialization.Tests.ReadJson<BinaryData>.DeserializeFromStream  |      1.02 |          1108.24 |          1135.76 |         |
| System.Text.Json.Serialization.Tests.ReadJson<Location>.DeserializeFromStream    |      1.02 |          1946.39 |          1994.25 |         |
| System.Text.Json.Serialization.Tests.ReadJson<ImmutableDictionary<String, String |      1.02 |         54959.14 |         56253.62 |         |
| System.Text.Json.Serialization.Tests.WriteJson<Hashtable>.SerializeToUtf8Bytes   |      1.02 |         21127.02 |         21617.07 |         |
| System.Text.Json.Serialization.Tests.ReadJson<SimpleStructWithProperties>.Deseri |      1.02 |           702.34 |           718.30 |         |

| Faster                                                                           | base/diff | Base Median (ns) | Diff Median (ns) | Modality|
| -------------------------------------------------------------------------------- | ---------:| ----------------:| ----------------:| --------:|
| System.Text.Json.Serialization.Tests.WriteJson<SimpleStructWithProperties>.Seria |      1.07 |           611.65 |           571.14 |         |
| System.Text.Json.Serialization.Tests.WriteJson<BinaryData>.SerializeToStream     |      1.05 |           634.36 |           603.74 |         |
| System.Text.Json.Serialization.Tests.WriteJson<IndexViewModel>.SerializeToStream |      1.05 |         22086.72 |         21041.37 |         |
| System.Text.Json.Serialization.Tests.WriteJson<SimpleStructWithProperties>.Seria |      1.04 |           362.26 |           349.30 |         |
| System.Text.Json.Serialization.Tests.WriteJson<SimpleStructWithProperties>.Seria |      1.04 |           436.05 |           420.82 |         |
| System.Text.Json.Serialization.Tests.WriteJson<Dictionary<String, String>>.Seria |      1.03 |         13726.78 |         13355.25 |         |
| System.Text.Json.Serialization.Tests.WriteJson<IndexViewModel>.SerializeToString |      1.02 |         23519.59 |         22983.97 |         |

layomia commented:
@Romanx, thanks for benchmarking and sharing these numbers.
The regressions here are concerning. I'll try to repro & take another pass at review for this PR.

Romanx commented:
Thanks very much, I'm not sure if it's my machine but I tried to keep a stable baseline during the tests and think I ran them correctly

layomia commented:
@Romanx, thank you for all the work and investigation you put into this PR. Unfortunately, the perf risk and complexity is too large at this stage of the release, given that the scenario being enabled is relatively minor given [other JSON feature requests ](https://github.com/dotnet/runtime/projects/25). We'll be willing to consider this in the future.

Romanx commented:
@layomia no worries at all, was glad to help at all.
